test/Pipeline.hs

```diff
@@ -0,0 +1,196 @@
+module Pipeline where
+
+import qualified Juvix.Backends.Michelson.Compilation as M
+import Juvix.Backends.Michelson.Compilation.Types
+import Juvix.Backends.Michelson.Parameterisation
+import qualified Juvix.Core.HR as HR
+import qualified Juvix.Core.IR as IR
+import qualified Juvix.Core.IR.Typechecker as Typed
+import qualified Juvix.Core.Pipeline as P
+import qualified Juvix.Core.Types as Core
+import qualified Juvix.Core.Usage as Usage
+import Juvix.Library hiding (bool, identity, log)
+import qualified Michelson.Typed as MT
+import qualified Michelson.Untyped as M
+import qualified Test.Tasty as T
+import qualified Test.Tasty.HUnit as T
+import Prelude (String)
+
+data Env primTy primVal
+  = Env
+      { parameterisation :: Core.Parameterisation primTy primVal,
+        log :: [Core.PipelineLog primTy primVal],
+        globals :: IR.Globals primTy primVal
+      }
+  deriving (Generic)
+
+type EnvExecAlias primTy primVal compErr =
+  ExceptT
+    (Core.PipelineError primTy primVal compErr)
+    (StateT (Env primTy primVal) IO)
+
+newtype EnvExec primTy primVal compErr a
+  = EnvE (EnvExecAlias primTy primVal compErr a)
+  deriving (Functor, Applicative, Monad, MonadIO)
+  deriving
+    ( HasSink "log" [Core.PipelineLog primTy primVal],
+      HasWriter "log" [Core.PipelineLog primTy primVal]
+    )
+    via WriterField "log" (EnvExecAlias primTy primVal compErr)
+  deriving
+    ( HasReader "parameterisation" (Core.Parameterisation primTy primVal),
+      HasSource "parameterisation" (Core.Parameterisation primTy primVal)
+    )
+    via ReaderField "parameterisation" (EnvExecAlias primTy primVal compErr)
+  deriving
+    ( HasState "globals" (IR.Globals primTy primVal),
+      HasSource "globals" (IR.Globals primTy primVal),
+      HasSink "globals" (IR.Globals primTy primVal)
+    )
+    via StateField "globals" (EnvExecAlias primTy primVal compErr)
+  deriving
+    (HasReader "globals" (IR.Globals primTy primVal))
+    via ReaderField "globals" (EnvExecAlias primTy primVal compErr)
+  deriving
+    (HasThrow "error" (Core.PipelineError primTy primVal compErr))
+    via MonadError (EnvExecAlias primTy primVal compErr)
```

Comment:
We could probably add this to the src

Reply:
yeah, but first we should unify the stack at least somewhat - https://github.com/metastatedev/juvix/issues/159 

right now, this is only in `app`, which we should avoid `tests` being dependent upon